### PR TITLE
feat: Add deck.gl Heatmap Visualization

### PR DESCRIPTION
**Pull request by Mattc1221**
### SUMMARY

Adds the heatmap layer from deck.gl [here](https://deck.gl/docs/api-reference/aggregation-layers/heatmap-layer).

Related SIP: https://github.com/apache/superset/discussions/23553

**_Rationale_**:

Superset already has supported functionality for other deck.gl visualization layers (e.g. Arc, Geojson, Grid, Hex, Path, Polygon, Scatter, Screengrid). Since Superset already contains the logic to handle deck.gl layers, it is straightforward and easy to maintain to add new deck.gl layers. 

By expanding the range of available deck.gl visualization layers, users will have more options to choose from when creating their visualizations. This will allow them to tailor their visualizations to their specific needs and explore their data in different ways. It will also allow Superset to better compete with other data visualization tools on the market that offer a wider range of visualization options.

Overall, I think adding more deck.gl layers to Superset is a great idea and has the potential to significantly improve the user experience and the insights that users can gain from their data.

**_Design Decisions_**

The addition of the Heatmap layer follows the structure of existing deck.gl layers to maintain readability and consistency within the codebase. 

Structure: 
The directory `superset-frontend/plugins/legacy-preset-chart-deckgl/src/layers/Heatmap/` contains the front end structure for the heat map
- `controlPanel` configures the form to grab relevant information specified by the user to create and tune the heatmap.
- `Heatmap.jsx` contains functions that define the Heatmap Layer, control the tooltip when hover the chart, and to get relevant data
- `index.js` serves as the wrapper for the heatmap component and provides a metadata function to grab information to descrive what this visualization is/does

The Back End code for this feature follows the back end code for other deck.gl visualizations as well, providing functions to structure queried data for the Heatmap component.

### BEFORE/AFTER SCREENSHOTS OR ANIMATED GIF

New visualization recording:

https://user-images.githubusercontent.com/36670322/229327601-22f81978-174e-41c0-8f8f-6cf7ea66691d.mov

### TESTING INSTRUCTIONS

1. Start Superset
2. Click on the `Charts` tab at the top of the screen
3. Click the `+ Chart` button at the top right
4. Load a dataset with geo-positional data (longitude/latitude)
5. Search the charts for `Deck.gl Heatmap` or click on the `Map` chart category
6. Select `deck.gl Heatmap` and hit `Create New Chart`
7. Update the `Longitude & Latitude` field to the correct columns
8. Select a `Weight` function
9. Select `Create Chart`
10. There should now be a heat map layer present on the screen
    - Note: the underlying map component may not be visible if the MAP_BOX_API_KEY is not set. To set it follow the instructions [here](https://superset.apache.org/docs/frequently-asked-questions/#why-is-the-map-not-visible-in-the-geospatial-visualization)
11. Change the options in the control panel, check that the information tooltips are correct/make sense for each field
12. Check that when hovering a point on the map, the latitude/longitude shows up correctly

### ADDITIONAL INFORMATION

- [ ] Has associated issue:
- [ ] Required feature flags:
- [ ] Changes UI
- [ ] Includes DB Migration (follow approval process in [SIP-59](https://github.com/apache/superset/issues/13351))
  - [ ] Migration is atomic, supports rollback & is backwards-compatible
  - [ ] Confirm DB migration upgrade and downgrade tested
  - [ ] Runtime estimates and downtime expectations provided
- [X] Introduces new feature or API
- [ ] Removes existing feature or API

### CODE REVIEW

**Comment by villebro** [comment]
This is really exciting work @Mattc1221 ! I'm marking myself as a reviewer and will do my best to review+test this later this week.

**Comment by Mattc1221** [comment]
Hi @villebro I believe this PR should now be ready for review! 

However, I noticed the tests have not run and I was wondering if I needed to close the PR and open a new one?

**Comment by Mattc1221** [comment]
@rusackas Thanks for re-running the tests :) 

The failing tests should now be fixed!

File: superset-frontend/plugins/legacy-preset-chart-deckgl/src/layers/Heatmap/index.js [hunk 1]
```diff
@@ -0,0 +1,45 @@
+/**
```
Comment (kgabryje) [inline superset-frontend/plugins/legacy-preset-chart-deckgl/src/layers/Heatmap/index.js:1]: Same as above - js -> ts 🙏

**Comment by kgabryje** [comment]
Awesome first contribution, congrats! I left a comment about refactoring js to ts, but otherwise the code looks great 🎉

**Comment by Mattc1221** [comment]
Thanks for the review @kgabryje 😀I'll update those JavaScript files to typescript sometime tomorrow evening!

**Comment by Mattc1221** [comment]
@kgabryje I really appreciate your first review! The changes requests should now be addressed and I would love to get another round of review!

**Review by kgabryje — APPROVED** [review]

**Comment by kgabryje** [comment]
@Mattc1221  Sorry for the long delay and thank you for implementing the changes 🙂 Will merge as soon as the CI passes

**Comment by kgabryje** [comment]
@Mattc1221 looks like there are some linting issues

**Comment by Mattc1221** [comment]
@kgabryje Thanks again for the review, and I apologize for the delayed response! I've rebased this branch to include the latest changes on the master branch and fixed the linting errors for the failing file `plugins/legacy-preset-chart-deckgl/types/external.d.ts`!

**Comment by Mattc1221** [comment]
Hey @kgabryje! could we run the CI tests again? With the latest changes, the tests pass locally and also on my GitHub actions page!


Screenshot 2023-05-20 at 2 18 16 PM

**Review by villebro — APPROVED** [review]
LGTM, really excited to see geospatial heatmaps arrive in Superset 🎉

**Comment by v1bg** [comment]
Hi! This is really cool. Just deployed out and tested and everything looks good with the new heatmap chart type, but just FYI when you try to add the heatmap to the "deck.gl Multiple Layers" chart type it doesn't appear to be working. Wanted to mention this to see if it's a known issue.

Thanks again!

**Comment by Mattc1221** [comment]
Big thanks to @kgabryje @villebro and @v1bg for reviewing this PR!

> Hi! This is really cool. Just deployed out and tested and everything looks good with the new heatmap chart type, but just FYI when you try to add the heatmap to the "deck.gl Multiple Layers" chart type it doesn't appear to be working. Wanted to mention this to see if it's a known issue.

This may have been an oversight and may be a missing import in the Multiple Layers chart type. I'll look into this further later this week and open a corresponding issue!

**Comment by lolvib** [comment]
> 

Appreciate it @Mattc1221! Great work here 🙏🏾

**Comment by v1bg** [comment]
Awesome, appreciate it @Mattc1221! Great work here 🙏🏾

**Comment by salvatorecolori** [comment]
Hi, does this plugin/chart have extraProps enabled for javascript tooltips generator? When I try to use js code to generate tooltpis, I receive a javascript Uncaught TypeError: Cannot read properties of undefined (reading 'extraProps')". 
Please check [https://github.com/apache/superset/issues/25871#issuecomment-2129426493](url) too.

**Pull request merged**